### PR TITLE
Use HIDE_SYMBOLS_BY_DEFAULT, fix MovingWindowFilter

### PR DESCRIPTION
**Pull request by j-rivero**
# 🦟 Bug fix

As part of the works of defining `hidden by default` in https://github.com/gazebosim/gz-cmake/pull/392, this PR moves the whole MovingWindowFilter class to be header only. Reason for this is that GCC on Jammy seems buggy to me (clang11 on Jammy worked fine, also clang on Mac and MSVC on Windows deal with it just fine) when it needs to generate the symbols needed for the specialization of the class done for `<gz::math::Vector3i>`.  See efforts in #565.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by azeey** [comment]
Since gz-cmake already has `HIDDEN_SYMBOLS_BY_DEFAULT`, maybe we should use that in all the libraries first, make sure all CI passes and then merge https://github.com/gazebosim/gz-cmake/pull/392. Thoughts?

**Comment by scpeters** [comment]
> Since gz-cmake already has `HIDDEN_SYMBOLS_BY_DEFAULT`, maybe we should use that in all the libraries first, make sure all CI passes and then merge [gazebosim/gz-cmake#392](https://github.com/gazebosim/gz-cmake/pull/392). Thoughts?

this sounds like you are suggesting the following:

1. Pass `HIDE_SYMBOLS_BY_DEFAULT` to the `gz_configure_build` calls in every `gz-*` package
2. Test and verify that CI / nightly builds are happy
3. Remove `HIDE_SYMBOLS_BY_DEFAULT` from the `gz_configure_build` calls in every `gz-*` package
4. Merge https://github.com/gazebosim/gz-cmake/pull/392

If we merge https://github.com/gazebosim/gz-cmake/pull/392 before removing the `HIDE_SYMBOLS_BY_DEFAULT` parameters, I think it may introduce cmake deprecation warnings. I guess if we split out the deprecation of `HIDE_SYMBOLS_BY_DEFAULT`, then we could do things in a slightly different order, but it would require an extra gz-cmake pull request

**Comment by scpeters** [comment]
> > Since gz-cmake already has `HIDDEN_SYMBOLS_BY_DEFAULT`, maybe we should use that in all the libraries first, make sure all CI passes and then merge [gazebosim/gz-cmake#392](https://github.com/gazebosim/gz-cmake/pull/392). Thoughts?
> 
> this sounds like you are suggesting the following:
> 
> 1. Pass `HIDE_SYMBOLS_BY_DEFAULT` to the `gz_configure_build` calls in every `gz-*` package
> 2. Test and verify that CI / nightly builds are happy
> 3. Remove `HIDE_SYMBOLS_BY_DEFAULT` from the `gz_configure_build` calls in every `gz-*` package
> 4. Merge [Use visibility hidden by default gz-cmake#392](https://github.com/gazebosim/gz-cmake/pull/392)
> 
> If we merge [gazebosim/gz-cmake#392](https://github.com/gazebosim/gz-cmake/pull/392) before removing the `HIDE_SYMBOLS_BY_DEFAULT` parameters, I think it may introduce cmake deprecation warnings. I guess if we split out the deprecation of `HIDE_SYMBOLS_BY_DEFAULT`, then we could do things in a slightly different order, but it would require an extra gz-cmake pull request

it sounds a little like overkill, but I think it would be the most thorough approach since it would run CI on every single package

**Comment by azeey** [comment]
> it sounds a little like overkill, but I think it would be the most thorough approach since it would run CI on every single package

Maybe a little overkill, but my gut feeling is that it won't be straightforward to get all the packages to build and test properly with symbols hidden by default, so I'd like CI to run on the PRs rather than just local testing. If we can do that without using the `HIDE_SYMBOLS_BY_DEFAULT` approach, I'm all for it.

**Comment by scpeters** [comment]
> > it sounds a little like overkill, but I think it would be the most thorough approach since it would run CI on every single package
> 
> Maybe a little overkill, but my gut feeling is that it won't be straightforward to get all the packages to build and test properly with symbols hidden by default, so I'd like CI to run on the PRs rather than just local testing. If we can do that without using the `HIDE_SYMBOLS_BY_DEFAULT` approach, I'm all for it.

I opened https://github.com/gazebosim/gz-utils/pull/115 for example. I think it will take some additional development from our infrastructure team to support testing all packages from source on Ubuntu in CI

File: include/gz/math/MovingWindowFilter.hh [hunk 1]
```diff
@@ -90,6 +90,127 @@ namespace gz
...
+    return T(value);
+  }
+  else
+  {
+    auto value = this->sum / static_cast<double>(this->samples);
+    return T(value);
+  }
+}
+
+template class MovingWindowFilter<int>;
```
Comment (azeey) [inline include/gz/math/MovingWindowFilter.hh:207]: Do we still need these if it's a header-only?

**Comment by scpeters** [comment]
I opened #567 to set `HIDE_SYMBOLS_BY_DEFAULT` confirmed that CI fails, then I merged the changes into this PR

File: include/gz/math/MovingWindowFilter.hh (hunk 1 above)
Comment (j-rivero) [reply include/gz/math/MovingWindowFilter.hh]: You are right, I don't think so. [56f145f](https://github.com/gazebosim/gz-math/pull/566/commits/56f145f1cd10ad34d5403c5f51c720057343f96c)

**Comment by azeey** [comment]
@j-rivero Looks like there are two homebrew jobs? And both of them are failing trying to install either `gz-math` or `ign-math8`, neither of which are available.

image

**Comment by scpeters** [comment]
> @j-rivero Looks like there are two homebrew jobs? And both of them are failing trying to install either `gz-math` or `ign-math8`, neither of which are available.
> 
> image

I think it's due to https://github.com/gazebo-tooling/release-tools/pull/1054

**Comment by scpeters** [comment]
> @j-rivero Looks like there are two homebrew jobs? And both of them are failing trying to install either `gz-math` or `ign-math8`, neither of which are available.
> 
> image

I pushed a whitespace change (https://github.com/gazebosim/gz-math/pull/566/commits/17970d787fc5e348c11c8da6a86a0ab2088f0ec1) in order to re-run CI, and it's happy now

**Comment by scpeters** [comment]
I'll wait for @azeey to approve

**Review by azeey — APPROVED** [review]

**Pull request merged**